Add support for a declaration attribute `@warn` for source-level warning group behavior control

artemcm opened the pull request:
This PR contains two changes/commits: 

- Add parsing for an experimental declaration attribute '@warn' for source-level warning group behavior control. This attribute is gated behind an experimental feature `SourceWarningControl`.

- Implement support for unified warning group behavior queries per-diagnostic-group:
  - Refactor handling of `-Wwarning` and `-Werror` flags, as well as `DefaultIgnore` diagnostic groups to be captured as a `WarningGroupBehaviorRule` (as opposed to a prior `WarningAsErrorRule`) and replace the corresponding diagnostic group ID-indexed bit vectors in `DiagnosticEngine` with with a dictionary keyed by the diagnostic group ID with values representing behavior specification rules (such as `AsError`, `AsWarning`, `Ignored`). 
  - Add the capability to query the `SwiftWarningControl` SwiftSyntax [library](https://github.com/swiftlang/swift-syntax/blob/main/Sources/SwiftWarningControl/SwiftWarningControl.md) for a given warning diagnostic's diagnostic group's emission behavior at the diagnostic's source location, as determined by syntactic `@warn` controls. (Implemented in `ASTGen/WarningGroupBehaviorConfiguration.swift`).
  - When emitting a warning belonging to a diagnostic group, the diagnostic engine now supplies the above warning handling rules to the `SwiftWarningControl` query as global control inputs. This means the same query will take into account both global warning control rules specified on the command-line, as well as source-location-specific containing `@warn` attribute controls. 
    - More-specifically, for a given `ExportedSourceFile`, the first query of `warningGroupBehaviorAtPosition` will result in construction of a `WarningControlRegionTree` [data-structure](https://github.com/swiftlang/swift-syntax/blob/02f8220bccfa03df68cc50631e11891c7ebcca0c/Sources/SwiftWarningControl/WarningControlRegions.swift#L102) which will be cached in the `ExportedSourceFile` it belongs to and re-used for future queries.

artemcm committed:
Add parsing for a declaration attribute '@warn' for source-level warning group behavior control

artemcm committed:
Implement support for unified warning group behavior queries per-diagnostic

Unified across module-wide configuration flags (`-Wwarning`, `-Werror`, etc.) and syntactic configuration attribute `@warn`.

artemcm commented:
@swift-ci test

DougGregor commented on lib/AST/Bridging/DeclAttributeBridging.cpp:
```diff
@@ -487,6 +487,39 @@ BridgedSwiftNativeObjCRuntimeBaseAttr_createParsed(BridgedASTContext cContext,
+  
+  std::optional<StringRef> reasonText = std::nullopt;
+  if (!reason.getIsEmpty())
+    reasonText = reason.unbridged();
```
Don't we need to copy this string into the ASTContext?

artemcm replied:
Good catch, thank you.

DougGregor reviewed (commented):
Love this :)

artemcm commented:
https://github.com/swiftlang/swift-syntax/pull/3179
@swift-ci test

artemcm commented:
@swift-ci test

artemcm commented:
@swift-ci test Windows platform

artemcm commented:
@swift-ci test

artemcm committed:
De-register temporary buffer 'SourceFile's created during availability attribute argument parsing

Otherwise these source file objects linger in the source manager but they refer to a fully-temporary 'ASTContext' which does not exist after parsing is complete, which means they cannot be used in any way and attempting to do so would lead to a crash.

xymus commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -1768,6 +1768,13 @@ WARNING(attr_warn_unused_result_removed,none,
+      
+// warn
+ERROR(attr_warn_expected_diagnostic_group_identifier,none,
+      "expected '%0' option to be a diagnostic group identifier", (StringRef))
```
Do we want this one to be an error or should it just be a warning? IIUC an error would make it source breaking to remove a diagnostic group and could break source backward compatibility when adding this attribute for a new diagnostic group.

In the same sense, if we want to be more clear about the diagnostic group being unknown by the current compiler we could use something like `the diagnostic group identifier '%0' is unknown`.

artemcm replied:
That's a great point, changed to a warning. Thank you.

xymus commented on lib/Serialization/ModuleFormat.h:
```diff
@@ -2342,6 +2342,11 @@ namespace decls_block {
+  using WarnDeclAttrLayout = BCRecordLayout<
+    Warn_DECL_ATTR,
+    BCFixed<1> // implicit flag
+  >;
```
This is a mostly safe addition but I'd bump up `SWIFTMODULE_VERSION_MINOR` if you happen to make further changes to this PR, especially if we want to cherry-pick this to other branches.

xymus reviewed (approved):
I think the change to `RestrictNonProductionExperimentalFeatures` needs to be dropped, besides that this LGTM!

artemcm commented:
@swift-ci test Windows platform
